core/server/worker/src/main/java/alluxio/worker/page/PagedBlockStore.java

```diff
@@ -214,51 +219,85 @@ public BlockReader createBlockReader(long sessionId, long blockId, long offset,
           unpinBlock(blockLock.get());
         });
       }
     }
     // this is a block that needs to be read from UFS
-    UfsBlockReadOptions readOptions = UfsBlockReadOptions.fromProto(options);
     try (LockResource lock = new LockResource(mPageMetaStore.getLock().writeLock())) {
       // in case someone else has added this block while we wait for the lock,
       // just use the block meta; otherwise create a new one and add to the metastore
-      PagedBlockMeta blockMeta = mPageMetaStore
-          .getBlock(blockId)
-          .orElseGet(() -> {
-            long blockSize = options.getBlockSize();
-            PagedBlockStoreDir dir =
-                (PagedBlockStoreDir) mPageMetaStore.allocate(String.valueOf(blockId), 0);
-            PagedBlockMeta newBlockMeta = new PagedBlockMeta(blockId, blockSize, dir);
-            mPageMetaStore.addBlock(newBlockMeta);
-            dir.getEvictor().addPinnedBlock(blockId);
-            return newBlockMeta;
-          });
-
-      final PagedBlockReader pagedBlockReader = new PagedBlockReader(mCacheManager,
-          mUfsManager, mUfsInStreamCache, mConf, blockMeta, offset, Optional.of(readOptions));
-      return new DelegatingBlockReader(pagedBlockReader, () -> {
-        blockMeta.getDir().getEvictor().removePinnedBlock(blockId);
+      Optional<PagedBlockMeta> blockMeta = mPageMetaStore.getBlock(blockId);
+      if (blockMeta.isPresent()) {
+        blockMeta.get().getDir().getEvictor().addPinnedBlock(blockId);
+        return new DelegatingBlockReader(getBlockReader(blockMeta.get(), offset, options), () -> {
+          blockMeta.get().getDir().getEvictor().removePinnedBlock(blockId);
+          unpinBlock(blockLock.get());
+        });
+      }
+      PagedBlockStoreDir dir =
+          (PagedBlockStoreDir) mPageMetaStore.allocate(String.valueOf(blockId),
+              options.getBlockSize());
+      PagedBlockMeta newBlockMeta = new PagedBlockMeta(blockId, options.getBlockSize(), dir);
+      if (options.getNoCache()) {
+        // block does not need to be cached in Alluxio, no need to add and commit it
+        unpinBlock(blockLock.get());
+        return new PagedUfsBlockReader(mUfsManager, mUfsInStreamCache, mConf, newBlockMeta,
+            offset, getUfsBlockReadOptions(blockId, options));
+      }
+      mPageMetaStore.addBlock(newBlockMeta);
+      dir.getEvictor().addPinnedBlock(blockId);
+      return new DelegatingBlockReader(getBlockReader(newBlockMeta, offset, options), () -> {
+        commitBlockToMaster(newBlockMeta);
+        newBlockMeta.getDir().getEvictor().removePinnedBlock(blockId);
         unpinBlock(blockLock.get());
       });
     }
   }
 
+  private BlockReader getBlockReader(PagedBlockMeta blockMeta, long offset,
+      Protocol.OpenUfsBlockOptions options) {
+    final long blockId = blockMeta.getBlockId();
+    final Optional<PagedUfsBlockReader> ufsBlockReader;
+    if (mPageMetaStore.hasFullBlock(blockId)) {
+      ufsBlockReader = Optional.empty();
```

Comment:
I understand you don't want to create the ufs reader for the fully cached block, but what if some page of this block got evicted during the read? 

I'm thinking if we could passed in a supplier of the ufsBlockReader, what do you think?

Reply:
I agree, I think the safer option here is we should try out best to provide a UFS reader even if the block is known to be fully cached. In case the client does not provide enough information about how to read the block from UFS, it's should be considered a bug in the client, and we should signal the error to the client clearly.

Reply:
done